File: tests/algorithms/x/Clojure/ciphers/gronsfeld_cipher.clj
(ns main (:refer-clojure :exclude [index_of to_uppercase gronsfeld]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare index_of to_uppercase gronsfeld)

(def ^:dynamic gronsfeld_ascii_len nil)

(def ^:dynamic gronsfeld_ch nil)

(def ^:dynamic gronsfeld_encrypted nil)

(def ^:dynamic gronsfeld_i nil)

(def ^:dynamic gronsfeld_idx nil)

(def ^:dynamic gronsfeld_key_idx nil)

(def ^:dynamic gronsfeld_key_len nil)

(def ^:dynamic gronsfeld_new_position nil)

(def ^:dynamic gronsfeld_shift nil)

(def ^:dynamic gronsfeld_upper_text nil)

(def ^:dynamic index_of_i nil)

(def ^:dynamic to_uppercase_ch nil)

(def ^:dynamic to_uppercase_i nil)

(def ^:dynamic to_uppercase_idx nil)

(def ^:dynamic to_uppercase_result nil)

(def ^:dynamic main_ASCII_UPPERCASE "ABCDEFGHIJKLMNOPQRSTUVWXYZ")

(def ^:dynamic main_ASCII_LOWERCASE "abcdefghijklmnopqrstuvwxyz")

(def ^:dynamic main_NEG_ONE (- 0 1))

(defn index_of [index_of_s index_of_ch]
  (binding [index_of_i nil] (try (do (set! index_of_i 0) (while (< index_of_i (count index_of_s)) (do (when (= (subs index_of_s index_of_i (min (+ index_of_i 1) (count index_of_s))) index_of_ch) (throw (ex-info "return" {:v index_of_i}))) (set! index_of_i (+ index_of_i 1)))) (throw (ex-info "return" {:v main_NEG_ONE}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn to_uppercase [to_uppercase_s]
  (binding [to_uppercase_ch nil to_uppercase_i nil to_uppercase_idx nil to_uppercase_result nil] (try (do (set! to_uppercase_result "") (set! to_uppercase_i 0) (while (< to_uppercase_i (count to_uppercase_s)) (do (set! to_uppercase_ch (subs to_uppercase_s to_uppercase_i (min (+ to_uppercase_i 1) (count to_uppercase_s)))) (set! to_uppercase_idx (index_of main_ASCII_LOWERCASE to_uppercase_ch)) (if (= to_uppercase_idx main_NEG_ONE) (set! to_uppercase_result (str to_uppercase_result to_uppercase_ch)) (set! to_uppercase_result (str to_uppercase_result (subs main_ASCII_UPPERCASE to_uppercase_idx (min (+ to_uppercase_idx 1) (count main_ASCII_UPPERCASE)))))) (set! to_uppercase_i (+ to_uppercase_i 1)))) (throw (ex-info "return" {:v to_uppercase_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn gronsfeld [gronsfeld_text gronsfeld_key]
  (binding [gronsfeld_ascii_len nil gronsfeld_ch nil gronsfeld_encrypted nil gronsfeld_i nil gronsfeld_idx nil gronsfeld_key_idx nil gronsfeld_key_len nil gronsfeld_new_position nil gronsfeld_shift nil gronsfeld_upper_text nil] (try (do (set! gronsfeld_ascii_len (count main_ASCII_UPPERCASE)) (set! gronsfeld_key_len (count gronsfeld_key)) (when (= gronsfeld_key_len 0) (throw (Exception. "integer modulo by zero"))) (set! gronsfeld_upper_text (to_uppercase gronsfeld_text)) (set! gronsfeld_encrypted "") (set! gronsfeld_i 0) (while (< gronsfeld_i (count gronsfeld_upper_text)) (do (set! gronsfeld_ch (subs gronsfeld_upper_text gronsfeld_i (min (+ gronsfeld_i 1) (count gronsfeld_upper_text)))) (set! gronsfeld_idx (index_of main_ASCII_UPPERCASE gronsfeld_ch)) (if (= gronsfeld_idx main_NEG_ONE) (set! gronsfeld_encrypted (str gronsfeld_encrypted gronsfeld_ch)) (do (set! gronsfeld_key_idx (mod gronsfeld_i gronsfeld_key_len)) (set! gronsfeld_shift (Integer/parseInt (subs gronsfeld_key gronsfeld_key_idx (min (+ gronsfeld_key_idx 1) (count gronsfeld_key))))) (set! gronsfeld_new_position (mod (+ gronsfeld_idx gronsfeld_shift) gronsfeld_ascii_len)) (set! gronsfeld_encrypted (str gronsfeld_encrypted (subs main_ASCII_UPPERCASE gronsfeld_new_position (min (+ gronsfeld_new_position 1) (count main_ASCII_UPPERCASE))))))) (set! gronsfeld_i (+ gronsfeld_i 1)))) (throw (ex-info "return" {:v gronsfeld_encrypted}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (gronsfeld "hello" "412"))
      (println (gronsfeld "hello" "123"))
      (println (gronsfeld "" "123"))
      (println (gronsfeld "yes, ¥€$ - _!@#%?" "0"))
      (println (gronsfeld "yes, ¥€$ - _!@#%?" "01"))
      (println (gronsfeld "yes, ¥€$ - _!@#%?" "012"))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Pascal/ciphers/gronsfeld_cipher.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  ASCII_UPPERCASE: string;
  ASCII_LOWERCASE: string;
  NEG_ONE: integer;
  s: string;
  ch: string;
  text: string;
  key: string;
function index_of(s: string; ch: string): integer; forward;
function to_uppercase(s: string): string; forward;
function gronsfeld(text: string; key: string): string; forward;
function index_of(s: string; ch: string): integer;
var
  index_of_i: integer;
begin
  index_of_i := 0;
  while index_of_i < Length(s) do begin
  if copy(s, index_of_i+1, (index_of_i + 1 - (index_of_i))) = ch then begin
  exit(index_of_i);
end;
  index_of_i := index_of_i + 1;
end;
  exit(NEG_ONE);
end;
function to_uppercase(s: string): string;
var
  to_uppercase_result_: string;
  to_uppercase_i: integer;
  to_uppercase_ch: string;
  to_uppercase_idx: integer;
begin
  to_uppercase_result_ := '';
  to_uppercase_i := 0;
  while to_uppercase_i < Length(s) do begin
  to_uppercase_ch := copy(s, to_uppercase_i+1, (to_uppercase_i + 1 - (to_uppercase_i)));
  to_uppercase_idx := index_of(ASCII_LOWERCASE, to_uppercase_ch);
  if to_uppercase_idx = NEG_ONE then begin
  to_uppercase_result_ := to_uppercase_result_ + to_uppercase_ch;
end else begin
  to_uppercase_result_ := to_uppercase_result_ + copy(ASCII_UPPERCASE, to_uppercase_idx+1, (to_uppercase_idx + 1 - (to_uppercase_idx)));
end;
  to_uppercase_i := to_uppercase_i + 1;
end;
  exit(to_uppercase_result_);
end;
function gronsfeld(text: string; key: string): string;
var
  gronsfeld_ascii_len: integer;
  gronsfeld_key_len: integer;
  gronsfeld_upper_text: string;
  gronsfeld_encrypted: string;
  gronsfeld_i: integer;
  gronsfeld_ch: string;
  gronsfeld_idx: integer;
  gronsfeld_key_idx: integer;
  gronsfeld_shift: integer;
  gronsfeld_new_position: integer;
begin
  gronsfeld_ascii_len := Length(ASCII_UPPERCASE);
  gronsfeld_key_len := Length(key);
  if gronsfeld_key_len = 0 then begin
  panic('integer modulo by zero');
end;
  gronsfeld_upper_text := to_uppercase(text);
  gronsfeld_encrypted := '';
  gronsfeld_i := 0;
  while gronsfeld_i < Length(gronsfeld_upper_text) do begin
  gronsfeld_ch := copy(gronsfeld_upper_text, gronsfeld_i+1, (gronsfeld_i + 1 - (gronsfeld_i)));
  gronsfeld_idx := index_of(ASCII_UPPERCASE, gronsfeld_ch);
  if gronsfeld_idx = NEG_ONE then begin
  gronsfeld_encrypted := gronsfeld_encrypted + gronsfeld_ch;
end else begin
  gronsfeld_key_idx := gronsfeld_i mod gronsfeld_key_len;
  gronsfeld_shift := StrToInt(copy(key, gronsfeld_key_idx+1, (gronsfeld_key_idx + 1 - (gronsfeld_key_idx))));
  gronsfeld_new_position := (gronsfeld_idx + gronsfeld_shift) mod gronsfeld_ascii_len;
  gronsfeld_encrypted := gronsfeld_encrypted + copy(ASCII_UPPERCASE, gronsfeld_new_position+1, (gronsfeld_new_position + 1 - (gronsfeld_new_position)));
end;
  gronsfeld_i := gronsfeld_i + 1;
end;
  exit(gronsfeld_encrypted);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  ASCII_UPPERCASE := 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  ASCII_LOWERCASE := 'abcdefghijklmnopqrstuvwxyz';
  NEG_ONE := 0 - 1;
  writeln(gronsfeld('hello', '412'));
  writeln(gronsfeld('hello', '123'));
  writeln(gronsfeld('', '123'));
  writeln(gronsfeld('yes, ¥€$ - _!@#%?', '0'));
  writeln(gronsfeld('yes, ¥€$ - _!@#%?', '01'));
  writeln(gronsfeld('yes, ¥€$ - _!@#%?', '012'));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
